shell_elvish.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"strings"
+)
+
+type elvish struct {
+	re *strings.Replacer
+}
+
+var ELVISH = elvish{re: strings.NewReplacer(
+	`"`, "\"",
+	"'", `\'`,
+)}
+
+func (elvish) Hook() (string, error) {
+	return `fn -direnv-hook {
+	direnv export elvish > .direnv.elv
+	-source .direnv.elv
+	/bin/rm .direnv.elv
```

Comment:
is it possible to eval the output directly?

Reply:
Okay, maybe add a note about the eval.

There are two issues with that implementation:
* two shells entering at the same time would overwrite the same file
* if the `-source` is failing the `.direnv.elv` would be staying around

To work around that issue, use `ioutil.TempFile("", "direnv.elv")` to allocate a new random filename. Because the tmpdir is usually tmpfs it will disappear on next reboot.

Reply:
My understanding of `direnv` is that to support a new shell (Elvish here), it needs to know two things:

* A piece of hook code to execute at shell startup (`direnv hook`)

* A way to serialize environment changes that can be understood by the shell (`direnv export`).

Now for most shells, the output of `direnv export` is a piece of shell code, but as I understand it, it doesn't have to be; the place where the output of `direnv export` used is in the hook code, and we control the producer as well as the consumer. For Elvish, it makes more sense to make this **not** shell code; evaluating arbitrary code is not well supported and generally frowned upon. Instead you can

* In `Export`, simply output `e` as a JSON map.

* In the hook function, use `from-json` to read the JSON map, iterate through it, and use `{set,unset}-env` to manipulate the environment (blame me for not documenting them; https://github.com/elves/elvish/issues/666).

Reply:
Brilliant. There is already a `direnv dump json` so all that needs to be done is on the elvish side. Most shells don't support JSON as an input so I didn't expect that :)